### ISSUE TITLE
Nicer syntax for higher level functions

### ISSUE THREAD

**Issue by dasch** [google/jsonnet#1002]
I love the pure functional nature of Jsonnet! However, when trying to do things in a functional style, the syntax gets in the way a _bit_ (not a lot, but a _bit_).

For example, defining higher level functions is a bit clumsy:

```jsonnet
local f = function(x) function(y) function(z) z*y*z;
```

I think such functions would be more natural to use, and therefore get used more often, with more succinct syntax – and there's no need to invent anything fancy, since lots of other languages already use either a "fat" or a "thin" _arrow_, `=>` or `->`.

So the above snippet would become:

```jsonnet
local f = x -> y -> z -> z*y*z;
```

This would also be nice and eligible for inline uses, e.g. `std.filter(thing -> thing.isActive, things)`.

Hoping y'all think it makes sense! 🤞

**Comment by sparkprime** [comment]
We probably want to use the backslash syntax for this because it's easy to parse but it's worth checking what other languages do and trying to be consistent with those if possible.

**Comment by CertainLach** [comment]
> Optional parentheses could still be used when necessary

I think it should be treated not as optional parentheses, but as optional parenthesis omission

JS allows it too in case of single identifier, i.e `(arg) => value` can be written as `arg => value`, but not for deconstruction/multiple arguments:
```js
([arr1, arr2]) => value
({field1, field2}) => value
(arg1, arg2) => value
```

Because it will make parsing much harder, and will require backtracing

**Comment by dasch** [comment]
@CertainLach yup, that makes sense. Again, my ML thinking leads me to think of "multiple arguments" as just a single value – a tuple. Indeed, one possible input to the deconstruction discussion would be to introduce the notion of tuples, with pattern matching allowing for setting defaults, e.g.

```jsonnet
local (a, b, c=42) = someFunc(x, y, z);

// Symmetrical with
local f = (a, b, c=42) -> ...;
f someFunc();
```

That last part would perhaps be a breaking change, though – not sure if we can clearly distinguish between the `a` argument to `f` being assigned the tuple vs. deconstructing.

**Comment by CertainLach** [comment]
Unfortunately, parameters in jsonnet isn't just a tuple deconstruction

They have non-trivial semantics regarding to named arguments (Jsonnet supports pythonic keyword arguments, `(function(a, b, c) null)(1, c = 3, b = 2)`) and bindings (Parameters may default to other parameters in arbitrary order, `function(a = c, b = a, c = b) null`)

**Comment by sbarzowski** [comment]
Related discussion:
https://github.com/google/jsonnet/issues/612
https://github.com/google/jsonnet/issues/223
